src/shogun/mathematics/graph/ops/abstract/ShogunStorage.h

```diff
@@ -0,0 +1,310 @@
+				};
+
+			public:
+				friend class ReshapeShogun;
+
+				ShogunStorage(
+				    const Shape& shape, const std::shared_ptr<NumberType>& type)
+				    : m_data(nullptr), m_shape(shape), m_type(type)
+				{
+					// shape is known at build time, let's preallocate it
+					// we could inforce immutability at this level somehow?
+					if (shape.is_static())
+						m_data = std::make_shared<InternalStorage>(
+						    get_size_from_shape(m_shape), type);
+				}
+
+				[[nodiscard]] std::shared_ptr<Tensor> to_tensor() const
+				{
+					auto tensor = std::make_shared<Tensor>(m_shape, m_type);
+					// transfer data to tensor
+					m_data->get_copy(
+					    tensor->data(), get_size_from_shape(m_shape), m_type);
+					tensor->m_free = true;
+					return tensor;
+				}
+
+				/* Equivalent as device_put in JAX. Transfers data to internal
+				 * storage.
+				 */
+				static std::shared_ptr<ShogunStorage>
+				from_tensor(const std::shared_ptr<Tensor>& tensor)
+				{
+					if (!tensor->get_shape().is_static())
+						error("Cannot copy a tensor of unknown shape.");
+					auto storage = std::make_shared<ShogunStorage>(
+					    tensor->get_shape(), tensor->get_type());
+					// transfer data from tensor
+					storage->m_data->copy_from(
+					    tensor->data(),
+					    get_size_from_shape(tensor->get_shape()),
+					    tensor->get_type());
+					return storage;
+				}
+
+				/* Allocates memory storage on the fly with a given shape.
+				 * If the exact is not static, e.g. the exact dimensions are
+				 * not known, at runtime, this will throw.
+				 *
+				 */
+				void allocate_storage(const Shape& shape)
+				{
+					if (!shape.is_static())
+						error(
+						    "Cannot allocate tensor with shape {}, with "
+						    "unknown size requirements",
+						    shape.to_string());
+					// new allocation, call allocator_dispatch
+					if (!m_data)
+					{
+						set_shape(shape);
+						m_data = std::make_shared<InternalStorage>(
+						    get_size_from_shape(shape), m_type);
+					}
+					else
+					{
+						// memory already allocated, and we have the right shape
+						// nothing to do
+						if (m_shape == shape)
+							return;
+						// memory has been allocated, but it isn't the same
+						// shape
+						else
+						{
+							auto old_shape = m_shape;
+							set_shape(shape);
+							// if the size requirement is larger, reallocate
+							// memory
+							if (get_size_from_shape(shape) > size())
+							{
+								m_data->realloc(
+								    get_size_from_shape(shape), m_type);
+							}
+							// otherwise nothing happens, we just own a larger
+							// memory block but only use part of it
+						}
+					}
+				}
+
+				[[nodiscard]] const Shape& get_shape() const noexcept
+				{
+					return m_shape;
+				}
+
+				[[nodiscard]] const std::shared_ptr<NumberType>&
+				get_type() const noexcept
+				{
+					return m_type;
+				}
+
+				[[nodiscard]] void* data() {
+					return m_data->m_internal_data.get();
```

Comment:
@vigsterkr do you know how this could be threadsafe?

Reply:
I mean could have a lock, but it would be nice to have all of that behind the scenes somehow